managers/bosh-manager/DirectorService.js

```diff
@@ -0,0 +1,1116 @@
+    return this.director.getDeploymentNames(queued);
+  }
+
+  getDeploymentNamesInCache() {
+    return boshOperationQueue.getDeploymentNames();
+  }
+
+  getTask(taskId) {
+    logger.info(`Fetching task '${taskId}'...`);
+    return this.director
+      .getTask(taskId)
+      .tap(task => logger.info(`+-> Fetched task for deployment '${task.deployment}' has state '${task.state}'`))
+      .catch(err => {
+        logger.error('+-> Failed to fetch task');
+        logger.error(err);
+        throw err;
+      });
+  }
+
+  getDeploymentManifest(deploymentName) {
+    logger.info(`Fetching deployment manifest '${deploymentName}'...`);
+    return this.director
+      .getDeploymentManifest(deploymentName)
+      .tap(() => logger.info('+-> Fetched deployment manifest'))
+      .catch(err => {
+        logger.error('+-> Failed to fetch deployment manifest');
+        logger.error(err);
+        throw err;
+      });
+  }
+
+  executePolicy(scheduled, action, deploymentName) {
+    const runOutput = {
+      'shouldRunNow': false
+    };
+    let targetDirectorConfig;
+    return this.director.getDirectorForOperation(action, deploymentName)
+      .then(directorConfig => {
+        targetDirectorConfig = directorConfig;
+        return this.director.getCurrentTasks(action, targetDirectorConfig);
+      })
+      .then(tasksCount => {
+        let currentTasks, maxWorkers;
+        const allTasks = tasksCount.total;
+        const maxTasks = _.get(targetDirectorConfig, 'max_workers', 6);
+        if (allTasks >= maxTasks) {
+          //no slots left anyway
+          runOutput.shouldRunNow = false;
+          return runOutput;
+        }
+        if (scheduled) {
+          currentTasks = tasksCount.scheduled;
+          maxWorkers = _.get(targetDirectorConfig, 'policies.scheduled.max_workers', 3);
+        } else {
+          currentTasks = tasksCount[action];
+          maxWorkers = _.get(targetDirectorConfig, `policies.user.${action}.max_workers`, 3);
+        }
+        if (currentTasks < maxWorkers) {
+          //should run if the tasks count is lesser than the specified max workers for op type
+          runOutput.shouldRunNow = true;
+          return runOutput;
+        }
+        return runOutput;
+      }).catch(err => {
+        logger.error('Error connecting to BOSH director > could not fetch current tasks', err);
+        //in case the director request returns an error, we queue it to avoid user experience issues
+        // return with shouldRunNow = false so that it is taken care of in processing
+        return runOutput;
+      });
+  }
+
+  _deleteEntity(action, opts) {
+    return utils.retry(tries => {
+        logger.info(`+-> Attempt ${tries + 1}, action "${opts.actionName}"...`);
+        return action();
+      }, {
+        maxAttempts: opts.maxAttempts,
+        minDelay: opts.minDelay
+      })
+      .catch(err => {
+        logger.error(`Timeout Error for action "${opts.actionName}" after multiple attempts`, err);
```

Comment:
How is it know that time out error occurred ?

Reply:
Can you please elaborate

Reply:
We are not checking type of error/exception here. Then how do we assume that it is a timeout error?